gitattributes.go

```diff
@@ -0,0 +1,251 @@
+func (gitAttrs *GitAttributes) IsDocumentation(path string) bool {
+	if val, ok := gitAttrs.boolAttributes[documentation].attributes[path]; ok {
+		return val
+	}
+
+	return data.DocumentationMatchers.Match(path)
+}
+
+// IsVendor returns whether or not path is a generated path.
+func (gitAttrs *GitAttributes) IsGenerated(path string) bool {
+	if val, ok := gitAttrs.boolAttributes[generated].attributes[path]; ok {
+		return val
+	}
+	return false
+}
+
+// GetLanguage get the language of a file matching the langauge attributes given.
+// Returns either a empty string or the language if the regExp matches
+func (gitAttrs *GitAttributes) GetLanguage(filename string) string {
+	for regExp, language := range gitAttrs.regExpAttributes[language].attributes {
+		if regExp.MatchString(filename) {
+			return language
+		}
+	}
+
+	return ""
+}
+
+// NewGitAttributes initialize a Gitattributes object
+func NewGitAttributes() *GitAttributes {
+	gitAttrs := GitAttributes{
+		boolAttributes: map[attrType]boolAttribute{
+			vendor:        boolAttribute{kind: vendor, matchers: []string{"linguist-vendored", "linguist-vendored=false"}, attributes: map[string]bool{}},
+			documentation: boolAttribute{kind: documentation, matchers: []string{"linguist-documentation", "linguist-documentation=false"}, attributes: map[string]bool{}},
+			generated:     boolAttribute{kind: generated, matchers: []string{"linguist-generated", "linguist-generated=false"}, attributes: map[string]bool{}},
+		},
+		regExpAttributes: map[attrType]regExpAttribute{
+			language: regExpAttribute{kind: language, matchers: []string{"linguist-language="}, attributes: map[*regexp.Regexp]string{}},
+		},
+	}
+
+	return &gitAttrs
+}
+
+// LoadGitattributes reads and parses the file .gitattributes which overrides the standard strategies
+// Returns slice of errors that have may ocurred in the load
+func (gitAttrs *GitAttributes) LoadGitAttributes(path string, reader io.Reader) []error {
+	rawAttributes, errArr := loadRawGitAttributes(reader)
+	if len(rawAttributes) == 0 {
+		return []error{}
+	}
+
+	return append(gitAttrs.parseAttributes(path, rawAttributes), errArr...)
+}
+
+func (gitAttrs *GitAttributes) String() string {
+	out := ""
+	for key, val := range gitAttrs.boolAttributes {
+		out += fmt.Sprintf("Type: %s	Attributes: %v\n", key, val.attributes)
+	}
+
+	for key, val := range gitAttrs.regExpAttributes {
+		out += fmt.Sprintf("Type: %s	Attributes: %v\n", key, val.attributes)
+	}
+	return out
+}
+
+func loadRawGitAttributes(reader io.Reader) (map[string][]string, []error) {
+	rawAttributes := map[string][]string{}
+	var errArr []error
+	data, err := ioutil.ReadAll(reader)
+	if err != nil {
+		errArr = append(errArr, err)
+		return nil, errArr
+	}
+
+	if len(data) > 0 {
+		lines := strings.Split(string(data), "\n")
+		for _, line := range lines {
+			err := loadLine(line, rawAttributes)
+			if err != nil {
+				errArr = append(errArr, err)
```

Comment:
If we found a malformed .gitattributes file, we can return a "malformed error" and return nil and one error.

I don't know if we should allow to load a malformed .gitattributes file and return a slice of errors. We can be sure that the rest of attributes are ok if there are errors in the file?